File: backend/handlers.go
package main

import (
	"fmt"
	"log"

	// "net/http"
	// "sync"

	"github.com/gorilla/websocket"
)

// Create a new game
func createGame(gameCode string, conn *websocket.Conn) {
	mutex.Lock()
	defer mutex.Unlock()

	if _, exists := games[gameCode]; exists {
		sendMessage(conn, "error", "Game code already exists")
		return
	}

	// Create a new game and add the host
	games[gameCode] = &Game{
		Host:    conn,
		Players: []*websocket.Conn{},
	}

	sendMessage(conn, "gameCreated", fmt.Sprintf("Game %s created successfully", gameCode))
	log.Printf("Game %s created by host", gameCode)
}

// Join an existing game
func joinGame(gameCode string, conn *websocket.Conn) {
	mutex.Lock()
	defer mutex.Unlock()

	game, exists := games[gameCode]
	if !exists {
		sendMessage(conn, "error", "Game code does not exist")
		return
	}

	// Add the player to the game's player list
	game.Players = append(game.Players, conn)
	sendMessage(conn, "gameJoined", fmt.Sprintf("Joined game %s", gameCode))
	sendMessage(game.Host, "playerJoined", "A new player has joined the game")

	log.Printf("Player joined game %s", gameCode)
}

// Start the game (only the host can do this)
func startGame(gameCode string, conn *websocket.Conn) {
	mutex.Lock()
	defer mutex.Unlock()

	game, exists := games[gameCode]
	if !exists || game.Host != conn {
		sendMessage(conn, "error", "Only the host can start the game")
		return
	}

	// Notify all players
	sendMessage(game.Host, "gameStarted", "Game has started")
	for _, player := range game.Players {
		sendMessage(player, "gameStarted", "Game has started")
	}

	log.Printf("Game %s started", gameCode)
}

// Close a game (only the host can do this)
func closeGame(gameCode string, conn *websocket.Conn) {
	mutex.Lock()
	defer mutex.Unlock()

	game, exists := games[gameCode]
	if !exists || game.Host != conn {
		sendMessage(conn, "error", "Only the host can close the game")
		return
	}

	// Notify all players
	for _, player := range game.Players {
		sendMessage(player, "gameClosed", "The game has been closed")
		player.Close()
	}

	// Remove the game from the map
	delete(games, gameCode)
	sendMessage(conn, "gameClosed", "Game has been closed")
	log.Printf("Game %s closed", gameCode)
}

File: backend/src/handlers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.handleCreateRoom = handleCreateRoom;
exports.handleJoinRoom = handleJoinRoom;
exports.handleStartRoom = handleStartRoom;
exports.handleCloseRoom = handleCloseRoom;
exports.handleExitRoom = handleExitRoom;
const types_1 = require("./types");
const handler_helpers_1 = require("./handler-helpers");
/**
 * Handles room creation.
 */
function handleCreateRoom(roomCode, callback, socket) {
    // Check that there is no room with the same code
    if ((0, handler_helpers_1.checkIfRoomExists)(roomCode, callback))
        return;
    // Check that user is not in any room
    if ((0, handler_helpers_1.checkIfInAnyRoom)(socket.id, callback))
        return;
    // Create the room
    types_1.rooms[roomCode] = {
        code: roomCode,
        host: socket.id,
        players: new Set([socket.id]),
    };
    socket.join(roomCode);
    console.log(`Room ${roomCode} created by ${socket.id}`);
    callback({ success: true });
}
/**
 * Handles joining a room.
 */
function handleJoinRoom(roomCode, callback, socket) {
    // Check that the room exists
    if ((0, handler_helpers_1.checkIfRoomDoesNotExist)(roomCode, callback))
        return;
    // Check that user is not already in any room
    if ((0, handler_helpers_1.checkIfInAnyRoom)(socket.id, callback))
        return;
    // Add the user to the room
    types_1.rooms[roomCode].players.add(socket.id);
    socket.join(roomCode);
    console.log(`User ${socket.id} joined room ${roomCode}`);
    callback({ success: true });
}
/**
 * Handles starting a room.
 */
function handleStartRoom(roomCode, callback, socket) {
    if ((0, handler_helpers_1.checkIfRoomDoesNotExist)(roomCode, callback))
        return;
    if ((0, handler_helpers_1.checkIfNotHost)(roomCode, callback, socket.id))
        return;
    socket.to(roomCode).emit("startGame");
    console.log(`Room ${roomCode} started by host ${socket.id}`);
    callback({ success: true });
}
/**
 * Handles closing a room.
 */
function handleCloseRoom(roomCode, callback, socket) {
    if ((0, handler_helpers_1.checkIfRoomDoesNotExist)(roomCode, callback))
        return;
    if ((0, handler_helpers_1.checkIfNotHost)(roomCode, callback, socket.id))
        return;
    delete types_1.rooms[roomCode];
    socket.to(roomCode).emit('exitRoom');
    socket.leave(roomCode);
    console.log(`Room ${roomCode} closed by ${socket.id}`);
    callback({ success: true });
}
/**
 * Handles exiting a room.
 */
function handleExitRoom(roomCode, callback, socket, roomIsClosed) {
    // If room is already closed by host, no need to check this
    if (!roomIsClosed) {
        // Check if there is a room to exit
        if ((0, handler_helpers_1.checkIfRoomDoesNotExist)(roomCode, callback))
            return;
    }
    // A user can only exit this room if it is a player of it
    if ((0, handler_helpers_1.checkIfInThisRoom)(roomCode, callback, socket.id))
        return;
    // Remove from rooms list as player
    types_1.rooms[roomCode].players.delete(socket.id);
    // Exit socket room
    socket.leave(roomCode);
    // Message
    console.log(`Room with code ${roomCode} exited by user: ${socket.id}`);
    callback({ success: true });
}
